go/arrow/compute/internal/kernels/vector_run_end_encode.go

```diff
@@ -0,0 +1,947 @@
+	}
+	return nil
+}
+
+type encoder interface {
+	CountNumberOfRuns() (numValid, numOutput int64)
+	PreallocOutput(*exec.KernelCtx, int64, *exec.ExecResult)
+	WriteEncodedRuns(*exec.ExecResult) int64
+}
+
+func runEndEncodeImpl[R RunEndsType](ctx *exec.KernelCtx, batch *exec.ExecSpan, out *exec.ExecResult) error {
+	// first pass: count the number of runs
+	var (
+		inputArr      = &batch.Values[0].Array
+		inputLen      = inputArr.Len
+		numOutputRuns int64
+		numValidRuns  int64
+		enc           encoder
+	)
+
+	if inputLen == 0 {
+		reeType := arrow.RunEndEncodedOf(exec.GetDataType[R](), inputArr.Type)
+		*out = exec.ExecResult{
+			Type: reeType,
+			Children: []exec.ArraySpan{
+				{Type: reeType.RunEnds()}, {Type: reeType.Encoded()},
+			},
+		}
+		return nil
+	}
+
+	if err := validateRunEndType[R](inputLen); err != nil {
+		return err
+	}
+
+	enc = newEncoder[R](inputArr)
+	numValidRuns, numOutputRuns = enc.CountNumberOfRuns()
+	enc.PreallocOutput(ctx, numOutputRuns, out)
+
+	out.Children[1].Nulls = numOutputRuns - numValidRuns
+
+	written := enc.WriteEncodedRuns(out)
+	debug.Assert(written == numOutputRuns, "mismatch number of written values")
+	return nil
+}
+
+func runEndEncodeExec(ctx *exec.KernelCtx, batch *exec.ExecSpan, out *exec.ExecResult) error {
+	reeType := ctx.State.(RunEndEncodeState).RunEndType
+	switch reeType.ID() {
+	case arrow.INT16:
+		return runEndEncodeImpl[int16](ctx, batch, out)
+	case arrow.INT32:
+		return runEndEncodeImpl[int32](ctx, batch, out)
+	case arrow.INT64:
+		return runEndEncodeImpl[int64](ctx, batch, out)
+	}
+
+	return fmt.Errorf("%w: bad run end type %s", arrow.ErrInvalid, reeType)
+}
+
+type decodeBool[R RunEndsType] struct {
+	inputLen, inputOffset int64
+	inputRunEnds          []R
+
+	inputPhysicalOffset int64
+	inputValidity       []byte
+	inputValues         []byte
+	inputValueOffset    int64
+}
+
+func (de *decodeBool[R]) PreallocOutput(ctx *exec.KernelCtx, out *exec.ExecResult) {
+	*out = exec.ExecResult{
+		Type: arrow.FixedWidthTypes.Boolean,
+		Len:  de.inputLen,
+	}
+
+	if len(de.inputValidity) != 0 {
+		out.Buffers[0].WrapBuffer(ctx.AllocateBitmap(de.inputLen))
+	}
+
+	out.Buffers[1].WrapBuffer(ctx.AllocateBitmap(de.inputLen))
+}
+
+func (de *decodeBool[R]) ExpandAllRuns(out *exec.ExecResult) int64 {
+	var (
+		writeOffset         int64
+		runLength, numValid int64
+		outputValues        = out.Buffers[1].Buf
+		prevRunEnd          = R(de.inputOffset)
+	)
+
+	for i, runEnd := range de.inputRunEnds[de.inputPhysicalOffset:] {
+		runLength, prevRunEnd = int64(runEnd-prevRunEnd), runEnd
+		// if this run is a null, clear the bits and update writeOffset
+		if len(de.inputValidity) != 0 && bitutil.BitIsNotSet(de.inputValidity, int(de.inputValueOffset+de.inputPhysicalOffset)+i) {
+			bitutil.SetBitsTo(out.Buffers[0].Buf, writeOffset, runLength, false)
+			writeOffset += runLength
+			continue
+		}
+
+		// if the output has a validity bitmap, update it with 1s
+		if len(out.Buffers[0].Buf) != 0 {
```

Comment:
This condition and `len(de.inputValidity)` could become a single boolean set outside the loop.

Reply:
fair point, we only set `out.Buffers[0].Buf` if `len(de.inputValidity) != 0`.